File: androidx/src/main/java/com/androidx/picker/MediaFolder.java
package com.androidx.picker;

import android.os.Parcel;
import android.os.Parcelable;

import java.io.Serializable;
import java.util.ArrayList;

/**
 * Created by didikee on 2017/12/7.
 */

public class MediaFolder implements Serializable, Parcelable {

    public String name;  //当前文件夹的名字
    public String path;  //当前文件夹的路径
    public ArrayList<MediaItem> items;  //当前文件夹下所有图片的集合

    public boolean check; //当前文件夹是否选中

    public MediaFolder() {
    }

    public MediaItem getCover() {
        if (items != null && items.size() > 0) {
            return items.get(0);
        }
        return null;
    }


    /** 只要文件夹的路径和名字相同，就认为是相同的文件夹 */
    @Override
    public boolean equals(Object o) {
        try {
            MediaFolder other = (MediaFolder) o;
            return this.path.equalsIgnoreCase(other.path) && this.name.equalsIgnoreCase(other.name);
        } catch (ClassCastException e) {
            e.printStackTrace();
        }
        return super.equals(o);
    }

    protected MediaFolder(Parcel in) {
        name = in.readString();
        path = in.readString();
        items = in.createTypedArrayList(MediaItem.CREATOR);
        check = in.readByte() != 0;
    }

    public static final Creator<MediaFolder> CREATOR = new Creator<MediaFolder>() {
        @Override
        public MediaFolder createFromParcel(Parcel in) {
            return new MediaFolder(in);
        }

        @Override
        public MediaFolder[] newArray(int size) {
            return new MediaFolder[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(name);
        dest.writeString(path);
        dest.writeTypedList(items);
        dest.writeByte((byte) (check ? 1 : 0));
    }
}

File: androidx/src/main/java/com/androidx/picker/MediaFolder.kt
package com.androidx.picker

import android.os.Parcel
import android.os.Parcelable
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize
import java.io.Serializable

/**
 * Created by didikee on 2017/12/7.
 */
@Parcelize
open class MediaFolder(
    @JvmField var name: String = "",  // 当前文件夹的名字
    @JvmField var path: String = "",  // 当前文件夹的路径
    @JvmField var items: ArrayList<MediaItem> = ArrayList(),  // 当前文件夹下所有图片的集合
    @JvmField var check: Boolean = false  // 当前文件夹是否选中
) : Parcelable {
    val id: String
        get() = "${path}/${name}"

    // 用于缓存计算的文件夹总大小，避免每次访问都进行计算
    @IgnoredOnParcel
    @Transient
    private var cachedTotalFileSize: Long = 0

    // 获取封面图片
    fun getCover(): MediaItem? {
        return if (items.isNotEmpty()) items[0] else null
    }

    // 获取文件夹大小
    fun getSize(): Int {
        return items.size
    }

    // 计算文件夹的总文件大小，并缓存
    fun getTotalFileSize(): Long {
        // 如果缓存的大小有效，直接返回缓存的值
        if (cachedTotalFileSize != 0L) {
            return cachedTotalFileSize
        }

        // 计算所有图片文件的大小
        cachedTotalFileSize = items.sumOf { it.size }
        return cachedTotalFileSize
    }


    // 只要文件夹的路径和名字相同，就认为是相同的文件夹
    override fun equals(other: Any?): Boolean {
        return if (other is MediaFolder) {
            this.path.equals(other.path, ignoreCase = true) && this.name.equals(
                other.name,
                ignoreCase = true
            )
        } else {
            super.equals(other)
        }
    }

    // 重写hashCode方法，保证与equals一致
    override fun hashCode(): Int {
        return 31 * name.hashCode() + path.hashCode()
    }
}
